[excel] (Power Automate) Show how to handle large datasets in Power Automate (#583)
OfficeDev/office-scripts-docs: docs/resources/samples/write-large-dataset.md

---
title: Write a large dataset
description: Learn how to split a large dataset into smaller write operations in Office Scripts.
ms.date: 02/24/2023
ms.localizationpriority: medium
---

# Write a large dataset

The `Range.setValues()` API puts data in a range. This API has limitations depending on various factors, such as data size and network settings. This means that if you attempt to write a massive amount of information to a workbook as a single operation, you'll need to write the data in smaller batches in order to reliably update a [large range](../../testing/platform-limits.md).

> The first part of the sample shows how to write a large dataset in Excel. The second part expands the example to be part of a Power Automate flow. This is necessary if your script takes longer to run that the [Power Automate action timeout](../../testing/platform-limits.md#power-automate).

Reviewer: There's a grammar issue between the sentence and the link text. Maybe this? 
```suggestion
The first part of the sample shows how to write a large dataset in Excel. The second part expands the example to be part of a Power Automate flow. This is necessary if your script takes long enough to run that the [Power Automate action times out](../../testing/platform-limits.md#power-automate).
```

}
```

### Training video: Write a large dataset

[Watch Sudhi Ramamurthy walk through this sample on YouTube](https://youtu.be/BP9Kp0Ltj7U).

## Sample 2: Write data in batches from a Power Automate flow

For this sample, you'll need to complete the following steps.

1. Create a workbook in OneDrive named **SampleData.xlsx**.
1. Create a second workbook in OneDrive named **TargetWorkbook.xlsx**.
1. Open  **SampleData.xlsx**.
1. Add sample data. You can use the script from the [Write a large dataset in batches](#sample-1-write-a-large-dataset-in-batches) section to generate this data.
1. Create and save both of the following scripts.
1. Follow the steps under [Power Automate flow: Read and write data in a loop](#power-automate-flow-read-and-write-data-in-a-loop) to create the flow.

### Sample code: Read part of a workbook

```TypeScript
function main(workbook: ExcelScript.Workbook, startRow: number, batchSize: number) : string[][] {
    // This sample only reads the first worksheet in the workbook.
    const sheet = workbook.getWorksheets()[0];

    // Get the boundaries of the range.

>     // Note that we're assuming usedRange is too big to read or write as a single Range.

Reviewer: Nit, for your consideration. 
```suggestion
    // Note that we're assuming usedRange is too big to read or write as a single range.
```

    return rangeToRead.getValues() as string[][];
}

```

### Sample code: Write part of a workbook

```TypeScript
function main(workbook: ExcelScript.Workbook, data: string[][], currentRow: number, batchSize: number): boolean {
  // Get the first worksheet.
  const sheet = workbook.getWorksheets()[0];

  // Set the given data.
  if (data && data.length > 0) {
    sheet.getRangeByIndexes(currentRow, 0, data.length, data[0].length).setValues(data);
  }

  // If we wrote less data than the batch size, signal the end of the flow.
  return batchSize > data.length;
}
```

### Power Automate flow: Read and write data in a loop

1. Sign into [Power Automate](https://flow.microsoft.com) and create a new **Instant cloud flow**.
1. Choose **Manually trigger a flow** and select **Create**.
1. Add a **New step** to track the current row being read and written. Make a new **Initialize variable** action with the following values.
    * **Name**: currentRow
    * **Type**: Integer
    * **Value**: 0

    :::image type="content" source="../../images/write-large-dataset-1.png" alt-text="The completed 'Initialize variable' step for the 'currentRow'.":::

> 1. Add a **New step** to set the number of rows to be read in a single batch. Depending on the number of columns, this made need to be smaller to avoid the data transfer limits. Make a new **Initialize variable** action with the following values.

Reviewer: Typo fix
```suggestion
1. Add a **New step** to set the number of rows to be read in a single batch. Depending on the number of columns, this may need to be smaller to avoid the data transfer limits. Make a new **Initialize variable** action with the following values.
```

    * **Name**: batchSize
    * **Type**: Integer
    * **Value**: 10000

    :::image type="content" source="../../images/write-large-dataset-2.png" alt-text="The completed 'Initialize variable' step for the 'batchSize'.":::
1. Add a **Do until** control. The flow will read chunks of the data until it has all been copied. You'll use the value of **-1** to indicate the end of the data has been reached. Give the control the following values.

>     * *First value*: *currentRow* (dynamic content)

Reviewer: For consistency with the other steps, I think First value, Condition, and Second value should be bolded instead of italicized in this section. Like this: 
```suggestion
    * **First value**: *currentRow* (dynamic content)
```

Author: I debated about this. The other bolded words refer to a specific label in the UX, whereas the first term "First value" is never actually displayed. Do you think I should still make it bold?

Author: That's a good compromise. That would mean that in these walkthroughs bold text is actual labels, italics are dynamic content, and regular text is conceptual spaces or user input.

Reviewer: I prefer the second option, I think it's clearer as a reader.

    * *Condition*: is equal to
    * *Second value*: -1

    :::image type="content" source="../../images/write-large-dataset-3.png" alt-text="The completed 'Do until' control.":::

> 1. The remaining steps are added inside the **Do until** control, call the script to read the data. Add an **Excel Online (Business)** connector with the **Run script** action. Use the following values for the action.

Reviewer: The flow is a little off here. Maybe a transition word? 
```suggestion
1. The remaining steps are added inside the **Do until** control. Next, call the script to read the data. Add an **Excel Online (Business)** connector with the **Run script** action. Use the following values for the action.
```

    :::image type="content" source="../../images/write-large-dataset-4.png" alt-text="The completed 'Run script' action for the script that reads the data.":::
1. Call the script to write the data. Add a second **Excel Online (Business)** connector with the **Run script** action. Use the following values for the action.
    * **Location**: OneDrive for Business
    * **Document Library**: OneDrive
    * **File**: "TargetWorkbook.xlsx" (as selected by the file picker)
    * **Script**: Write data at row location
    * **startRow**: *currentRow* (dynamic content)
    * **batchSize**: *batchSize* (dynamic content)

    :::image type="content" source="../../images/write-large-dataset-5.png" alt-text="The completed 'Run script' action for the script that writes the data.":::
1. Update the current row to reflect that a batch of data has been read and written. Add an **Increment variable** action with the following values.
    * **Name**: currentRow
    * **Value**: *batchSize* (dynamic content)

    :::image type="content" source="../../images/write-large-dataset-6.png" alt-text="The completed 'Initialize variable' step for the 'batchSize'.":::

> 1. Add a **Condition** control to check if the scripts have read everything. The "Write data at row location" script returns true when it has written fewer rows than the batch size allows. This means it's all the end of the data set. Create the **Condition** control with the following values.

Reviewer: I think this is a typo -- 
```suggestion
1. Add a **Condition** control to check if the scripts have read everything. The "Write data at row location" script returns true when it has written fewer rows than the batch size allows. This means it's at the end of the data set. Create the **Condition** control with the following values.
```

>     * *First value*: *result* (dynamic content from **Run script**)

Reviewer: Same comment as the previous italicized labels.
